### ISSUE TITLE
[map] Allow merging two maps

### ISSUE THREAD

**Issue by ldionne** [boostorg/hana#299]
Hana does not provide a way to merge two maps out of the box, although that would be very useful. For now, one has to use `fold_left` with `insert` to insert each element of one map into the other:

``` c++
#include <boost/hana.hpp>
#include <string>
namespace hana = boost::hana;
using namespace hana::literals;

constexpr auto m1 = hana::make_map(
    hana::make_pair("key1"_s, hana::type_c<std::string>),
    hana::make_pair("key2"_s, hana::type_c<std::string>)
);

constexpr auto m2 = hana::make_map(
    hana::make_pair("key3"_s, hana::type_c<std::string>),
    hana::make_pair("key4"_s, hana::type_c<std::string>),
    hana::make_pair("key5"_s, hana::type_c<std::string>)
);

constexpr auto result = hana::fold_left(m1, m2, hana::insert);
constexpr auto expected = hana::make_map(
    hana::make_pair("key1"_s, hana::type_c<std::string>),
    hana::make_pair("key2"_s, hana::type_c<std::string>),
    hana::make_pair("key3"_s, hana::type_c<std::string>),
    hana::make_pair("key4"_s, hana::type_c<std::string>),
    hana::make_pair("key5"_s, hana::type_c<std::string>)
);

static_assert(result == expected, "");

int main() { }
```

**Comment by ricejasonf** [comment]
I would expect that if both `m1` and `m2` had "key1" set to different values, the value in `m2` would prevail which is not the case with `hana::insert`.

**Comment by shreyans800755** [comment]
@ricejasonf @ldionne 
I would like to work on this feature. What kind of interface are we looking for ?
Are we looking for something like, hana::merge_map(m1, m2) or operator like + and += on two maps ?

**Comment by ricejasonf** [comment]
`hana::map` could possibly be implemented as a `hana::Monoid`. I don't think `+=` would make sense as it suggests you are modifying the first map.

**Comment by ldionne** [comment]
@shreyans800755 Thanks for your interest! I think we could reuse `hana::union_`, since a "merge" is really a union of two maps. So it would be a matter of

1. Specialize `hana::union_` for `map_tag`
2. Document that `hana::union_` works for maps too
3. Add tests for `hana::union_` on maps

**Comment by shreyans800755** [comment]
@ldionne it perfectly makes sense. I'll try to complete it asap.

**Comment by ldionne** [comment]
Take it easy! Feel free to submit an incomplete to gather early feedback if you want, too.

**Comment by ldionne** [comment]
Closed by #338. Thanks @shreyans800755! Note that we should also probably add a `hana::map` version of other set-theoretic operations.

**Comment by shreyans800755** [comment]
@ldionne I can work on that. May be I'll create an issue first and we can discuss it over gitter.
